src/api/client.ts

```diff
@@ -212,6 +214,33 @@ export class WsClient extends Emittery {
         sendHandler(request, resolve, reject);
       });
       return promise as Promise<Response>;
+    } else if (this.url && this.authenticationToken) {
+      const response = new Promise<unknown>((resolve, reject) => {
+        // typescript forgets in this promise scope that `this.url` is not undefined
+        const socket = new IsoWebSocket(this.url as URL, this.options);
+        this.socket = socket;
+        socket.onerror = (errorEvent) => {
```

Comment:
Can this tell the difference between an invalid token and other kinds of error?

It seems like the user experience wouldn't be great if requests always result in a reconnect that fails when a temporary token was in use. If we can tell the token is invalid here, it would be good to clear it so that the client stops trying to reconnect

Reply:
With an invalid token what happens is that you get the "Client disconnected with pending requests" error and this onError callback here is never called in the first place.

Reply:
I don't actually understand where the "client disconnected with pending requests" occurs...but if I log something in that `onError` callback it's not getting logged when it fails due to an invalid token.

Reply:
Interesting. Well Holochain won't send back anything to tell the client that the connection is going to be closed, it'll just close the connection from its end. Where is that error thrown, can/should it be caught?

Reply:
I confirm that his happens. It comes in as a `close` event on the websocket with code [`1005`](https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent/code#value) which means "No status received". It's indistinguishable from another unexpected close event which usually is 1005 or 1006. For this reconnect we could register another close handler that unsets the token, but it's not 100 % certain that the close was due to an invalid token.

Reply:
There's a specific `InvalidTokenError` now which only happens on failed automatic websocket reconnect. If the reconnect fails, the stored token is deleted. No risk of this happening in a loop, because the reconnects are only attempted as part of a request being made by the client consumer.